plugins/backstage-plugin-coder/src/api/CoderClient.ts

```diff
@@ -200,37 +200,39 @@ export class CoderClient implements CoderClientApi {
       request: WorkspacesRequest,
       config: CoderWorkspacesConfig,
     ): Promise<WorkspacesResponse> => {
-      const { workspaces } = await baseSdk.getWorkspaces(request);
-      const paramResults = await Promise.allSettled(
-        workspaces.map(ws =>
-          this.sdk.getWorkspaceBuildParameters(ws.latest_build.id),
-        ),
+      if (config.repoUrl === undefined) {
+        return { workspaces: [], count: 0 };
+      }
+
+      // Have to store value here so that type information doesn't degrade
+      // back to (string | undefined) inside the .map callback
+      const stringUrl = config.repoUrl;
+      const responses = await Promise.allSettled(
+        config.repoUrlParamKeys.map(key => {
+          const patchedRequest = {
+            ...request,
+            q: appendParamToQuery(request.q, key, stringUrl),
```

Comment:
are `key` and `stringUrl` guaranteed to be strings as opposed to null or undefined?

Reply:
Luckily, they're both guaranteed to be defined

With `key`, we're getting the values through `.map`. That doesn't guarantee that we'll have a non-empty array, but it does mean that the mapper callback will only be called if we have a defined element to feed into it. That's one of the benefits of the array methods and `for/of` over more loops with `i` – they can guarantee no null-ish values, because they're going through JS's iterator protocol

And as for `stringUrl`, it's wonky, but that's also guaranteed to be of type `string`. If we had this:
```ts
const patchedRequest = {
  ...request,
  q: appendParamToQuery(request.q, key, config.repoUrl),
};
```
That would be type `string | undefined`. But TS type narrowing helps get around that:

```ts
// Rule out that config.repoUrl is not undefined; TS is smart
// enough to shrink size down to just string
if (config.repoUrl === undefined) {
  return { workspaces: [], count: 0 };
}

// Capture the value after we've proven it's just a string
const stringUrl = config.repoUrl;

// Normally, the type of a union value would "reset" when you
// have a different callback, because TS can't look at the code and
// make guarantees about whether the value will change But as long
// as we use the "captured" version of the value, it will always be a string
const responses = /* */
```

Reply:
> But as long as we use the "captured" version of the value, it will always be a string and can't be reset

I see your comment above now. Thank you for the explanation!